src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8410,6 +8429,487 @@ namespace WaterThermalTanks {
+                            const Real64 Q_AdiabaticMixing = (Tmixed - Tfinal[k]) / FinalFactorMixing;
+                            Tfinal[k] = Tmixed;
+                            Tavg[k] += Q_AdiabaticMixing * AvgFactorMixing;
+                        }
+
+                        // Since we mixed, get out of here and start from the top to check again for mixing.
+                        break;
+                    }
+                }
+            } while ( HasInversion );
+
+            // Venting
+            if (Tfinal[0] > Tank.TankTempLimit) {
+                for (int i = 0; i < nTankNodes; ++i) {
+                    if (Tfinal[i] > Tank.TankTempLimit) {
+                        Event += Tank.Node[i].Mass * Cp * (Tank.TankTempLimit - Tfinal[i]);
+                        Tfinal[i] = Tank.TankTempLimit;
+                    }
+                }
+            }
+
+            // Increment to next internal time step
+            TimeRemaining -= dt;
+            Real64 Qloss = 0.0;
+            for (int i = 0; i < nTankNodes; ++i) {
+                auto &node = Tank.Node[i];
+                node.Temp = Tfinal[i];
+                node.TempSum += Tavg[i] * dt;
+
+                // Bookkeeping for reporting variables
+                if (Tank.HeaterOn1 || Tank.HeaterOn2) {
+                    Qloss += node.OnCycLossCoeff * (Tank.AmbientTemp - Tavg[i]);
+                } else {
+                    Qloss += node.OffCycLossCoeff * (Tank.AmbientTemp - Tavg[i]);
+                }
+            }
+            // More bookkeeping for reporting variables
+            Eloss += Qloss * dt;
+            const Real64 Quse = Tank.UseEffectiveness * Tank.UseMassFlowRate * Cp * (Tank.UseInletTemp - Tavg[Tank.UseOutletStratNode - 1]);
+            Euse += Quse * dt;
+            const Real64 Qsource = [&]{
+                if (Tank.HeatPumpNum > 0) {
+                    if (HPWHCondenserConfig == TypeOf_HeatPumpWtrHeaterPumped) {
+                        return Qheatpump;
+                    } else {
+                        assert(HPWHCondenserConfig == TypeOf_HeatPumpWtrHeaterWrapped);
+                        return 0.0;
+                    }
+                } else {
+                    return Tank.SourceEffectiveness * Tank.SourceMassFlowRate * Cp * (Tank.SourceInletTemp - Tavg[Tank.SourceOutletStratNode - 1]);
+                }
+            }();
+            Esource += Qsource * dt;
+            if (Tank.HeaterOn1) Runtime1 += dt;
+            if (Tank.HeaterOn2) Runtime2 += dt;
+            if (Tank.HeaterOn1 || Tank.HeaterOn2) Runtime += dt;
+            Eheater1 += Qheater1 * dt;
+            Eheater2 += Qheater2 * dt;
+            Eunmet += max(Tank.UseEffectiveness * Tank.UseMassFlowRate * (UseOutletExpectedTemp - Tavg[Tank.UseOutletStratNode - 1]), 0.0) * dt;
+
+        } // end while TimeRemaining > 0.0
+
+        for (auto &e : Tank.Node) {
+            e.TempAvg = e.TempSum / SecInTimeStep;
+            e.TempSum = 0.0;
+        }
```

Comment:
And isn't a += here or some average method of multiple nodes needed at 8831 if the loop is kept? Otherwise the last node is the average?

Reply:
Or is e.TempAvg an array?

Reply:
TempAvg is needed after here but this could still be calculated within the loop at 8792 to get rid of this loop.

Reply:
You're right I could write it that way. I did it like this because initially I expected `dt` to be a variable timestep. In the future we may end up going that way, so I'd like to leave it as is. I don't expect it's adding a noticeable number of clock cycles.